File: python/ai/predict_camera.py
#!/usr/bin/env python3
import json
import os
import sys
import joblib
import pandas as pd

def predict_speed_camera(model, latitude, longitude, time_of_day, day_of_week):
    if ":" in time_of_day:
        time_of_day = "evening" if int(time_of_day.split(":")[0]) > 18 else "morning"
    input_data = {
        'latitude': [latitude],
        'longitude': [longitude],
        'time_of_day': [time_of_day],
        'day_of_week': [day_of_week]
    }
    input_df = pd.DataFrame(input_data)
    input_df = pd.get_dummies(input_df, columns=['time_of_day', 'day_of_week'])
    for col in model.feature_names_in_:
        if col not in input_df.columns:
            input_df[col] = 0
    input_df = input_df[model.feature_names_in_]
    prediction = model.predict(input_df)
    return prediction[0].tolist()

def main():
    if len(sys.argv) != 5:
        print("[]")
        return 1
    latitude = float(sys.argv[1])
    longitude = float(sys.argv[2])
    time_of_day = sys.argv[3]
    day_of_week = sys.argv[4]
    model_path = os.path.join(os.path.dirname(__file__), "speed_camera_model.pkl")
    model = joblib.load(model_path)
    prediction = predict_speed_camera(model, latitude, longitude, time_of_day, day_of_week)
    print(json.dumps(prediction))
    return 0

if __name__ == "__main__":
    sys.exit(main())

File: python/ai/predict_camera.dart
#!/usr/bin/env dart
import 'dart:convert';
import 'dart:io';
import 'dart:math';
import 'package:path/path.dart' as p;

void main(List<String> args) async {
  if (args.length != 4) {
    stdout.writeln('[]');
    exit(1);
  }
  final latitude = double.parse(args[0]);
  final longitude = double.parse(args[1]);
  final timeOfDay = args[2];
  final dayOfWeek = args[3];
  final model = PredictiveModel();
  final result = predictSpeedCamera(
    model: model,
    latitude: latitude,
    longitude: longitude,
    timeOfDay: timeOfDay,
    dayOfWeek: dayOfWeek,
  );
  if (result == null) {
    stdout.writeln('[]');
    exit(1);
  } else {
    stdout.writeln(jsonEncode([result[0], result[1]]));
  }
}

class PredictiveModel {
  final List<Point<double>> cameras;
  PredictiveModel() : cameras = _load();

  static List<Point<double>> _load() {
    final scriptDir = p.dirname(Platform.script.toFilePath());
    final trainingPath = p.join(scriptDir, 'training.json');
    final data = jsonDecode(File(trainingPath).readAsStringSync())
        as Map<String, dynamic>;
    final cams = <Point<double>>[];
    for (final cam in data['cameras'] as List<dynamic>) {
      final coords = cam['coordinates'] as List<dynamic>;
      for (final coord in coords) {
        cams.add(Point<double>(
          (coord['latitude'] as num).toDouble(),
          (coord['longitude'] as num).toDouble(),
        ));
      }
    }
    return cams;
  }
}

List<double>? predictSpeedCamera({
  required PredictiveModel model,
  required double latitude,
  required double longitude,
  required String timeOfDay,
  required String dayOfWeek,
}) {
  if (model.cameras.isEmpty) return null;
  final current = Point<double>(latitude, longitude);
  Point<double>? best;
  var bestDist = double.infinity;
  for (final cam in model.cameras) {
    final dx = current.x - cam.x;
    final dy = current.y - cam.y;
    final dist = sqrt(dx * dx + dy * dy);
    if (dist < bestDist) {
      bestDist = dist;
      best = cam;
    }
  }
  if (best == null) return null;
  return [best.x, best.y];
}
